contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1024 @@
+   * Requirements:
+   *
+   * - the caller must have the `PAUSER_ROLE`.
+   */
+  function pause() external onlyRole(PAUSER_ROLE) {
+    _pause();
+  }
+
+  /**
+   * @dev Unpauses all token transfers.
+   *
+   * See {ERC1155Pausable} and {Pausable-_unpause}.
+   *
+   * Requirements:
+   *
+   * - the caller must have the `PAUSER_ROLE`.
+   */
+  function unpause() external onlyRole(PAUSER_ROLE) {
+    _unpause();
+  }
+
+  /**
+   * Registers the addresses of the market, bpNori, and removal contracts in this contract.
+   *
+   * ##### Requirements:
+   *
+   * - Can only be used when the caller has the `DEFAULT_ADMIN_ROLE`
+   */
+  function registerContractAddresses(
+    address marketAddress,
+    address bridgedPolygonNoriAddress,
+    address removalAddress
+  ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+    _market = FIFOMarket(marketAddress);
+    _bridgedPolygonNori = BridgedPolygonNORI(bridgedPolygonNoriAddress);
+    _removal = Removal(removalAddress);
+  }
+
+  /**
+   * Sets the duration in seconds that should be applied to schedules created on behalf of removals
+   * originating from the given methodology and methodology version.
+
+   * ##### Requirements:
+   *
+   * - Can only be used when the contract is not paused.
+   * - Can only be used when the caller has the `DEFAULT_ADMIN_ROLE`
+   */
+  function setRestrictionDurationForMethodologyAndVersion(
+    uint256 methodology,
+    uint256 methodologyVersion,
+    uint256 durationInSeconds
+  ) public whenNotPaused onlyRole(DEFAULT_ADMIN_ROLE) {
+    _methodologyAndVersionToScheduleDuration[methodology][
+      methodologyVersion
+    ] = RestrictionDuration(durationInSeconds, true);
+  }
+
+  /**
+   * Sets up a restriction schedule with parameters determined from the project id.
+   *
+   *
+   * ##### Requirements:
+   * - Can only be used when the contract is not paused.
+   * - Can only be used when the caller has the `SCHEDULE_CREATOR_ROLE` role
+   */
+  function createSchedule(uint256 projectId)
+    external
+    whenNotPaused
+    onlyRole(SCHEDULE_CREATOR_ROLE)
+  {
+    if (!_scheduleIdToScheduleStruct[projectId].exists) {
+      _createSchedule(projectId);
+    }
+  }
+
+  /**
+   * This function is triggered when BridgedPolygonNORI is sent to this contract
+   *
+   * @dev Sending BridgedPolygonNORI to this contract triggers the tokensReceived hook defined by the ERC-777 standard
+   * because this contract is a registered ERC777 tokens recipient.
+   *
+   * [See here for more](
+   * https://github.com/ethereum/EIPs/blob/master/EIPS/eip-777.md#erc777tokensrecipient-and-the-tokensreceived-hook)
+   */
+  function tokensReceived(
+    address,
+    address from,
+    address,
+    uint256 amount,
+    bytes calldata userData,
+    bytes calldata operatorData
+  ) external override {
+    if (!(_msgSender() == address(_bridgedPolygonNori))) {
+      revert TokenSenderNotBPNORI();
+    }
+    if (!(from == address(_market) || hasRole(DEFAULT_ADMIN_ROLE, from))) {
```

Comment:
I'd like to discuss this. Will paste more in slack.

Reply:
modified as discussed.  this also involves giving the `admin` the necessary depositor role _in the local setupTest function only_ for these tests, because they rely on being able to receive tokens directly for test setup purposes instead of the overkill of having to "sell" valid supply to seed each test